Core/include/Acts/Geometry/Portal.hpp

```diff
@@ -0,0 +1,166 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2022 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Definitions/Common.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/Geometry/GeometryIdentifier.hpp"
+#include "Acts/Geometry/NavigationDelegates.hpp"
+#include "Acts/Geometry/NavigationState.hpp"
+#include "Acts/Surfaces/BoundaryCheck.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+
+#include <map>
+#include <memory>
+#include <optional>
+
+namespace Acts {
+
+class ISurfaceMaterial;
+
+namespace Experimental {
+
+/// A portal description between the detector volumes
+///
+/// It has a Surface representation for navigation and propagation
+/// and guides from one volume to the next.
+///
+/// The surface can carry material to allow mapping onto
+/// portal positions if required.
+///
+class Portal : public std::enable_shared_from_this<Portal> {
+ protected:
+  /// Constructor from surface w/o portal links
+  ///
+  /// @param surface is the representing surface
+  Portal(std::shared_ptr<Surface> surface);
+
+ public:
+  /// The volume links forward/backward with respect to the surface normal
+  using DetectorVolumeUpdators = std::array<ManagedDetectorVolumeUpdator, 2u>;
+
+  /// The vector of attached volumes forward/backward, this is useful in the
+  /// geometry building
+  using AttachedDetectorVolumes =
+      std::array<std::vector<std::shared_ptr<DetectorVolume>>, 2u>;
+
+  /// Declare the DetectorVolume friend for portal setting
+  friend class DetectorVolume;
+
+  /// Factory for producing memory managed instances of Portal.
+  /// Will forward all parameters and will attempt to find a suitable
+  /// constructor.
+  template <typename... Args>
+  static std::shared_ptr<Portal> makeShared(Args&&... args) {
+    return std::shared_ptr<Portal>(new Portal(std::forward<Args>(args)...));
+  }
+
+  /// Retrieve a @c std::shared_ptr for this surface (non-const version)
+  ///
+  /// @note Will error if this was not created through the @c makeShared factory
+  ///       since it needs access to the original reference. In C++14 this is
+  ///       undefined behavior (but most likely implemented as a @c bad_weak_ptr
+  ///       exception), in C++17 it is defined as that exception.
+  /// @note Only call this if you need shared ownership of this object.
+  ///
+  /// @return The shared pointer
+  std::shared_ptr<Portal> getSharedPtr();
```

Comment:
In principle I think two `surface` accessors returning references should be enough, since we have have the `getSharedPtr` method for the surfaces as well, right?
I would prefer such an interface because it immediately tells the caller that the surface cannot be `nullptr` via interface.

Reply:
Agreed, changed.